starboard/shared/libiamf/iamf_audio_decoder.cc

```diff
@@ -0,0 +1,391 @@
+    const scoped_refptr<InputBuffer>& input_buffer) {
+  SB_DCHECK(BelongsToCurrentThread());
+  SB_DCHECK(input_buffer);
+  SB_DCHECK(output_cb_);
+  SB_DCHECK(!stream_ended_ || !pending_audio_buffers_.empty());
+  SB_DCHECK(is_valid());
+
+  if (input_buffer->size() == 0) {
+    SB_LOG(ERROR) << "Empty input buffer written to IamfAudioDecoder";
+    return false;
+  }
+
+  IamfConfigReader reader(input_buffer, kForceBinauralAudio,
+                          kForceSurroundAudio);
+  if (!reader.is_valid()) {
+    ReportError("Failed to parse IA Descriptors");
+    return false;
+  }
+  if (!decoder_is_configured_) {
+    if (!ConfigureDecoder(&reader, input_buffer->timestamp())) {
+      return false;
+    }
+  }
+
+  scoped_refptr<DecodedAudio> decoded_audio = new DecodedAudio(
+      audio_stream_info_.number_of_channels, GetSampleType(),
+      kSbMediaAudioFrameStorageTypeInterleaved, input_buffer->timestamp(),
+      audio_stream_info_.number_of_channels * reader.samples_per_buffer() *
+          starboard::media::GetBytesPerSample(GetSampleType()));
+  int samples_decoded = IAMF_decoder_decode(
+      decoder_, reader.data().data(), reader.data().size(), nullptr,
+      reinterpret_cast<void*>(decoded_audio->data()));
+  if (samples_decoded < 1) {
+    ReportError("Failed to decode IAMF sample, error " +
+                ErrorCodeToString(samples_decoded));
+    return false;
+  }
+
+  SB_DCHECK(samples_decoded <= reader.samples_per_buffer());
+
+  decoded_audio->ShrinkTo(audio_stream_info_.number_of_channels *
+                          reader.samples_per_buffer() *
+                          starboard::media::GetBytesPerSample(GetSampleType()));
+
+  // TODO: Enable partial audio once float32 pcm output is available.
+  const auto& sample_info = input_buffer->audio_sample_info();
+  decoded_audio->AdjustForDiscardedDurations(
+      audio_stream_info_.samples_per_second,
+      sample_info.discarded_duration_from_front,
+      sample_info.discarded_duration_from_back);
+
+  decoded_audios_.push(decoded_audio);
+
+  output_cb_();
+
+  return true;
+}
+
+void IamfAudioDecoder::WriteEndOfStream() {
+  SB_DCHECK(BelongsToCurrentThread());
+  SB_DCHECK(output_cb_);
+
+  stream_ended_ = true;
+  if (!pending_audio_buffers_.empty()) {
+    return;
+  }
+
+  // Put EOS into the queue.
+  decoded_audios_.push(new DecodedAudio);
+
+  Schedule(output_cb_);
+}
+
+bool IamfAudioDecoder::ConfigureDecoder(IamfConfigReader* reader,
+                                        int64_t timestamp) {
+  SB_DCHECK(is_valid());
+  SB_DCHECK(!decoder_is_configured_);
+
+  // TODO: libiamf has an issue outputting 32 bit float samples, set to 16 bit
+  // for now.
+  int error = IAMF_decoder_set_bit_depth(decoder_, 16);
+  if (error != IAMF_OK) {
+    ReportError("IAMF_decoder_set_bit_depth() fails with error " +
+                ErrorCodeToString(error));
+    return false;
+  }
+
+  error = IAMF_decoder_set_sampling_rate(decoder_, kDefaultSampleRate);
```

Comment:
Shall we always set the output to 48000? Is there any downside for doing that?

Reply:
In the near term, Youtube will only have IAMF videos with Opus substreams, which are always 48000. In the long term, we should set it based on the qualities of the stream - AAC substreams may be 441000 instead.